unitex/src/fr/umlv/unitex/tfst/tagging/TaggingModel.java

```diff
@@ -105,27 +150,677 @@ protected void updateModel() {
+	 * Put a path in "current"
+	 * When the path is complete, put it in allPaths
+	 * Then we restart
+	 * 
+	 * @param bfp 
+	 * 			The first box in the path
+	 * 
+	 * @param bfs
+	 * 			The last box in the path
+	 * 
+	 * @param current
+	 * 			The list of boxes of the current path
+	 * 
+	 * @param allPath
+	 * 			The list of all paths between the box bfp and bfs
+	 */	
+	void computePath(GenericGraphBox box, GenericGraphBox last, ArrayList<Integer> currentPath, ArrayList<ArrayList<Integer>> allPaths) {
+		if(box.equals(last))
+			allPaths.add(currentPath);
+		else {
+			currentPath.add(renumber[box.getBoxNumber()]);
+			for (GenericGraphBox next : box.transitions) {
+				computePath(boxes[next.getBoxNumber()], last, new ArrayList<Integer>(currentPath), allPaths);
+			}
+			
+		}
+	}
+	
+	void clearList() {
+		lstTok.clear();
+		lstPath.clear();
+	}
+	
+	void createLstContext() {
+		lstContext.clear();
+		for (String arrayList : lstPath) 
+			lstContext.add(new Context(arrayList));
+		
+		for (Context c : lstContext) {
+			for (int i = 0; i < c.path.size(); i++) {
+				if(getTextBoxe(boxes[sortedNodes[c.path.get(i)]]).equals("<E>")
+						&& (boxes[sortedNodes[c.path.get(i)]]).type != GenericGraphBox.INITIAL
+					&& (boxes[sortedNodes[c.path.get(i)]]).type != GenericGraphBox.FINAL) {
+					if(c.currentBox > i)
+						c.currentBox--;
+					c.path.remove(i);
+				}
+			}
+		}
+	}
+	
+	/** This function checks if the new branch can be add to the automaton or not
+	 * If it's not, we remove the last transition in the automaton
+	 */		
+	void checkNewBranch( int i ){
+		int prev = getPreviousFactorizationNodeIndex(i);
+		int next = getNextFactorizationNodeIndex(i);
+		ArrayList<ArrayList<Integer>> allPaths = new ArrayList<>();		
+		
+		computeAllPaths( renumber[prev], renumber[next], allPaths );
+		clearList();
+		findString(boxes[sortedNodes[renumber[prev]]], boxes[sortedNodes[renumber[next]]], new StringBuilder(), boxes[sortedNodes[renumber[prev]]], new StringBuilder());
+		createLstContext();
+		
+		int cpt = 0;
+		for (Context c : lstContext) {
+			cpt++;
+		}
+		
+		if(cpt == 0) {
+			return;
+		}
+		
+		TaggingState[] taggingStatesTmp = new TaggingState[taggingStates.length];
+		copyStates(taggingStates, taggingStatesTmp);
+		
+		
+		boolean b = verifyAllPath(allPaths, sortedNodes[renumber[prev]], sortedNodes[renumber[next]]);
+		int n = zone.graphBoxes.size();
+		for (int i1 = 0; i1 < n; i1++) {
+			boxes[i1] = (TfstGraphBox) zone.graphBoxes.get(i1);
+			if (boxes[i1].type == GenericGraphBox.INITIAL) {
+				initialState = i1;
+			}
+			else if (boxes[i1].type == GenericGraphBox.FINAL)
+				finalState = i1;
+			taggingStates[i1] = TaggingState.NEUTRAL;
+			boxes[i].state = TaggingState.NEUTRAL;
+		}
+		
+		
+		if(b == false)  {
+			zone.removeLastTransition();
+			copyStates(taggingStatesTmp, taggingStates);
 		}
-		 updateNodes();
+
+		updateNodes();
+		
+	}
+	
+	private void copyStates(TaggingState[] src, TaggingState[] dst) {
```

Comment:
This method does not use first parameter and does not copy; calls to this method suggest the expected behaviour of the method is to copy first parameter into second parameter